src/thorin/transform/inliner.cpp

```diff
@@ -53,8 +53,16 @@ void inliner(World& world) {
     auto is_candidate = [&] (Continuation* continuation) -> Scope* {
         if (!continuation->empty() && continuation->order() > 1) {
             auto scope = get_scope(continuation);
-            if (scope->defs().size() < scope->entry()->num_params() * factor + offset)
+            if (scope->defs().size() < scope->entry()->num_params() * factor + offset) {
+                // check that the function is not recursive to prevent inliner from peeling loops
+                for (auto& use : continuation->uses()) {
+                    // note that there was an edge from parameter to continuation,
+                    // we would need to check if the use is a parameter here.
+                    if (scope->contains(use.def()))
```

Comment:
I guess you meant: `note if there were an edge from parameter to continuation`

Reply:
Correct.